[CLIPPER-137] [CLIPPER-155] [CLIPPER-158] Deploy Scala Spark models no repl support

dcrankshaw opened the pull request:
Waiting on a few things:

- [x] Tests
- [ ] Documentation
- [x] #121 will create massive conflicts with this PR, so let's wait to even review until that PR is merged

Creating the pull request so I get a nice diff and can start to review it myself.

dcrankshaw committed:
# This is a combination of 6 commits.
# This is the 1st commit message:
started work on spark container

# This is the commit message #2:

rearranged files in maven packages

# This is the commit message #3:

training example works

# This is the commit message #4:

Copies jar to known location

# This is the commit message #5:

loading container class from jar not working yet

# This is the commit message #6:

got jar loading working

dcrankshaw committed:
initial work on incorporating REPL classes.

some cleanup and re-org. untested

basic example working

loading classes defined in REPL still doesn't work

started java reorg

Reorganized code. Everything compiles

Finished implementation. Now need docker container and lots of debugging.

start work on docker container

fixed bug in docker file

Fixed some bugs

Clipper local deploy seems to be working

Fixed some bugs

Deploy is working but there's something wrong with the container

Remote deploy is working!

Spark scala deploy is working both locally and remotely

Formatting

started to add tests

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/198/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/199/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/234/
Test FAILed.

dcrankshaw committed:
mostly done with rebase

dcrankshaw committed:
rebase complete with existing tests passing

dcrankshaw committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/259/
Test FAILed.

dcrankshaw committed:
Fix typo

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/260/
Test PASSed.

Corey-Zumar commented on containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/container/NoOpStringModel.java:
```diff
@@ -0,0 +1,30 @@
+import ai.clipper.container.data.FloatVector;
+import ai.clipper.container.data.SerializableString;
+
+import javax.xml.soap.SAAJResult;
```
Clean up imports. The following are unused:

```
import javax.xml.soap.SAAJResult;
import java.nio.FloatBuffer;
import java.util.Iterator;
import ai.clipper.container.data.FloatVector;
```

Corey-Zumar commented on containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala:
```diff
@@ -0,0 +1,108 @@
+package ai.clipper.examples.train
+
+
+// import scala.reflect.runtime.universe._
```
Remove commented import

Corey-Zumar commented on containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala:
```diff
@@ -0,0 +1,108 @@
+
+class LogisticRegressionContainer extends MLlibContainer {
+
+  // var model: Option[MLlibLogisticRegressionModel] = None
```
Clean commented code

dcrankshaw commented:
To test this:

1. Build example spark job: `cd <clipper-root>/containers/jvm && mvn clean package`
2. Download [Spark 2.1](http://spark.apache.org/downloads.html)
3. Build the Spark Container docker image: `cd <clipper-root>/containers/jvm && docker build -t clipper/spark-scala-container -f SparkScalaContainerDockerfile ./`
4. Start Clipper
5. Run the example Spark job using Spark submit : `CLIPPER_MODEL_NAME=<model-name> CLIPPER_HOST=<host> <path-to-spark>/bin/spark-submit --master "local[2]" --class ai.clipper.examples.train.Train --name <spark-job-name> <path-to-clipper>/containers/jvm/clipper-examples/target/clipper-examples-0.1.jar`

Corey-Zumar commented on containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala:
```diff
@@ -0,0 +1,108 @@
+      (point.label, prediction)
+    }
+    val numWrong = labelAndPreds.filter(r => r._1 != r._2).count()
+    println(s"Num wrong: $numWrong")
```
Let's print the test error as a percentage like we do with the Python examples - seems more meaningful than a quantity of incorrect predictions

Corey-Zumar commented on containers/jvm/clipper-rpc/src/main/java/ai/clipper/rpc/RPC.java:
```diff
@@ -227,23 +227,21 @@ private void serveModel(Model<I> model, final ZMQ.Context context, String ip, in
-    }
+  private void handlePredictRequest(long msgId, Iterator<I> dataVectors, ClipperModel<I> model,
+      ZMQ.Socket socket) throws IOException {
+    ArrayList<I> inputs = new ArrayList<I>();
```
Very nit: `ArrayList<I> inputs = new ArrayList<>();` suffices

Corey-Zumar commented on containers/jvm/clipper-rpc/src/test/scala/ai/clipper/rpc/RPCEventHistoryTest.scala:
```diff
@@ -0,0 +1,130 @@
+import ai.clipper.rpc.logging.{RPCEventHistory, RPCEventType}
+import org.scalatest.FunSuite
+//
+//import clipper.container.app.logging.RPCEvent;
```
Clean up imports.

Corey-Zumar commented on containers/jvm/clipper-rpc/src/test/scala/ai/clipper/rpc/RPCEventHistoryTest.scala:
```diff
@@ -0,0 +1,130 @@
+
+
+
+//public class RPCEventHistoryTest {
```
Remove commented, obsolete JUnit tests

Corey-Zumar commented on containers/jvm/clipper-spark/pom.xml:
```diff
@@ -0,0 +1,109 @@
+    <!--     <artifactId>asm</artifactId> -->
+    <!-- </dependency> -->
+    <!-- <dependency> -->
+    <!--   <groupId>org.apache.bcel</groupId> -->
```
Not sure if commented pom was auto-generated and or will be necessary later. If not, let's remove it.

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+  val CONTAINER_JAR_FILE: String = "container_source.jar"
+  val MODEL_DIRECTORY: String = "model"
+  val REPL_CLASS_DIR: String = "repl_classes"
+  // TODO: Uncomment before committing
```
Uncomment this before committing! i.e.

`dcrankshaw/spark-scala-container` --> `clipper/spark-scala-container`

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+  val CLIPPER_MANAGEMENT_PORT: Int = 1338
+
+
+  implicit val json4sFormats = Serialization.formats(NoTypeHints) + ModelTypeSerializer
```
Doc this! What is it for?

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+  implicit val json4sFormats = Serialization.formats(NoTypeHints) + ModelTypeSerializer
+
+
+  def deploySparkModel[M](sc: SparkContext,
```
When documenting, be sure to include the requirement that model inherit from `MLlibModel` or `PipelineModel`

dcrankshaw replied:
Added method documentation to this method that specifies this.

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+      "run",
+      "-d",
+      s"--network=$DOCKER_NW",
+      // "-m", "8g",
```
Remove

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+
+    try {
+      saveSparkModel[M](sc, name, version, model, containerClass, basePath)
+      publishModelToClipper(clipperHost, name, version, labels, hostDataPath)
```
We should call `publishModelToClipper` after determining whether or not `clipperHost` is local. If it is not local, we should check for the presence of ssh information before attempting to publish. We shouldn't publish a model that can't be deployed due to an absence of credentials.

dcrankshaw replied:
Done. Moved it below the check for sudo credentials and below copying the model data to Clipper. We will only publish the model to Clipper if the ssh credentials were both supplied and have access to the machine.

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+    val modelType = model match {
+      case m: MLlibModel => {
+        m.save(sc, modelPath)
+        // Because I'm not sure how to do it in the type system, check that
```
The first part of the comment indicating uncertainty seems unnecessary

dcrankshaw replied:
I actually want to leave it in there, because I'm pretty sure there is a way to do it with the type system, which would be preferable. If someone (including future me) comes along and has a bright idea about how to do that, that comment may prompt them to do something.

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+    val conf =
+      ClipperContainerConf(containerClass.getName, copiedJarName, modelType)
+    getReplOutputDir(sc) match {
+      case Some(classSourceDir) => {
```
We will never hit this case due to the check against `getReplOutputDir(sc).isDefined` above

dcrankshaw replied:
That's true, but Scala wants an exhaustive pattern match. And I'm leaving some commented out code in there intentionally because I hope to support deploying from the REPL in the future.

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -0,0 +1,346 @@
+      }
+      case PipelineModelType => {
+        val model = PipelineModel.load(modelPath)
+        val container = classLoader
```
Consider defining a `createSparkContainer` method to avoid repeating code for each case

dcrankshaw replied:
I think that the additional type parameter wrangling will actually make that less readable

Corey-Zumar commented on containers/jvm/clipper-spark/src/test/scala/ai/clipper/spark/ClipperTest.scala:
```diff
@@ -0,0 +1,346 @@
+package ai.clipper.spark
```
This file is a duplicate of `containers/jvm/clipper-spark/src/test/scala/ai/clipper/spark/Clipper.scala`

dcrankshaw replied:
Ahh good catch

Corey-Zumar reviewed (changes requested):
Most comments pertain to code cleanup (removing extraneous commented code, etc). Several small changes requested.

dcrankshaw committed:
added unit tests

dcrankshaw committed:
addressed review comments

dcrankshaw commented:
@Corey-Zumar I addressed the review comments. Take another look if you feel like it but hold off on merging for now. I'd like to do a bit more testing and add a little more documentation tomorrow before we merge.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/277/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/278/
Test PASSed.

dcrankshaw committed:
add some notes on how to run Spark example

dcrankshaw commented:
@Corey-Zumar Okay this is ready to go I think

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/282/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/283/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/284/
Test PASSed.

Corey-Zumar commented on containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:
```diff
@@ -52,17 +51,41 @@ object Clipper {
+    * @param version The model version
+    * @param model The trained Spark model. Note that this _must_ be an instance of either
+    *              ai.clipper.spark.MLlibModel or org.apache.spark.ml.PipelineModel
+    * @param containerClass This model container which specifies how to use the trained
```
"The model container" instead of "This model container"

Corey-Zumar reviewed (changes requested):
Looks good! Just one small documentation change to avoid potential confusion

dcrankshaw committed:
fixed typo in comment

dcrankshaw commented:
Addressed

Corey-Zumar reviewed (approved):
LGTM
